Shader Execution Reordering (SER) proposal (#277)
microsoft/hlsl-specs: proposals/0025-shader-execution-reordering.md

following sequence:

```C++
HitObject Hit = HitObject::TraceRay( ..., Payload );
ReorderThread( Hit );
HitObject::Invoke( Hit, Payload );
```

This snippet traces a ray and stores the result of traversal, intersection
testing and anyhit shading in `Hit`. The call to `ReorderThread` improves
coherence based on the information inside the `Hit`. Closesthit or miss
shading is then invoked in a more coherent context.

Note that this is a very basic example. Among other things, it is possible to
query information about the hit to influence `ReorderThread` with additional
hints. See [Separation of ReorderThread and HitObject::Invoke](#separation-of-reorderthread-and-hitobjectinvoke)
for more elaborate examples.

### HitObject HLSL Additions

```C++
HitObject
```

The `HitObject` type encapsulates information about a hit or a miss. A
`HitObject` is constructed using `HitObject::TraceRay`,
`HitObject::FromRayQuery`, `HitObject::MakeMiss`, or `HitObject::MakeNop`. It
can be used to invoke closesthit or miss shading using `HitObject::Invoke`,
and to reorder threads for shading coherence with `ReorderThread`.

The `HitObject` has value semantics, so modifying one `HitObject` will not

> impact any other `HitObject` in the shader. A shader can have any number of

Reviewer: Is there a reason for HitObject objects having value semantics while RayQuery objects have reference semantics?

Reviewer 2: In design meeting we agreed to review DXIL spec and revisit.

active `HitObject`s at a time. Each `HitObject` will take up some hardware
resources to hold information related to the hit or miss, including ray,

> intersection attributes and information about the shader to invoke. The size
> of a `HitObject` is unspecified. As a consequence, `HitObject` cannot be
> stored in structured buffers. Ray payload structs must not have members of
> `HitObject` type. `HitObject` supports assignment (by-value copy) and can be
> passed as arguments to and returned from local functions.

Reviewer 3: Description of restriction should mention that the type is `intangible`, and specify more locations that are disallowed.

```suggestion
intersection attributes and information about the shader to invoke. The size
of a `HitObject` is unspecified, and thus considered intangible. As a
consequence, `HitObject` cannot be stored in memory buffers, in groupshared
memory, in ray payloads, or in intersection attributes. `HitObject` supports
assignment (by-value copy) and can be passed as arguments to and returned from
local inlined functions.
```

A `HitObject` is default-initialized to encode a NOP-HitObject (see `HitObject::MakeNop`).
A NOP-HitObject can be used with `HitObject::Invoke` and `ReorderThread` but

> does not call shaders or provide additional information for reordering.

Reviewer 3: > does not call shaders or provide additional information for reordering.

Should it be specified that both operations do nothing (`ReorderThread` causes no reordering)?  Or is it valid for an implementation to do thread reordering here when `HitObject` is a NOP-HitObject?

Reviewer 4: I read it as that it is valid to do reordering, though no hint or any other info are available. So, likely `ReorderThread` does no reordering. But we don’t want to *guarantee* that it does nothing for the same reason that we don’t want to guarantee reconvergence with a `[NoReorder]` attribute.

Author: It is valid to reorder even with a NOP-HitObject. At the very least this makes sense if not all threads in a wave have a NOP-HitObject. Hints can still be used as well.

Reviewer 3: @rasmusnv, I think this is a good explanation, thanks.

account when defining the ordering. For example, an implementation may decide
to order threads with respect to their hit group index, hit locations in
3d-space, or other factors.

`ReorderThread` may access both information about the instance in the
acceleration structure as well as the shader record at the shader table
offset contained in the `HitObject`. The respective fields in the `HitObject`
must therefore represent valid instances and shader table offsets.
NOP-HitObjects is an exception, which do not contain information about a hit
or a miss, but are still legal inputs to `ReorderThread`. Similarly, a
`HitObject` constructed from a `RayQuery` but did not set a shader table
index is exempt from having a valid shader table record.

```C++
void ReorderThread( HitObject Hit );
```

Parameter                           | Definition
---------                           | ----------
`HitObject Hit` | `HitObject` that encapsulates the hit or miss according to which reordering should be performed.

---

#### ReorderThread with coherence hint

This variant of `ReorderThread` reorders threads based on a generic
user-provided hint. Similarity of hint values should indicate expected

> similarity of subsequent work being performed by threads. The resolution of

Reviewer 3: ```suggestion
similarity of subsequent work being performed by threads. In general, more
significant bits of the hint value are more important than less significant
bits for determining coherency, though scheduling behavior is implementation
specific. The resolution of
```

the hint is implementation-specific. If an implementation cannot resolve all
values of `CoherenceHint`, it is free to ignore an arbitrary number of least

> significant bits. The thread ordering resulting from this call may be

Reviewer 5: We'd like to have a clearer description of how hit is mapping to coherence.
Could we add the sentence that higher bit of hint has more impact on coherence then lower ones.

e.g. consider 3 hints, and chose the most coherent pair of hint:
`A = 0b100000`
`B = 0b111111`
`C = 0b011111`

For example 3 methods for extracting coherence by defining opposite divergence metric:
1. If higher bit has more coherence then all lower bits after it, `A`,`B` pair is most coherent. Divergence more or less xor(...)
2. If its arithmetic difference then `A` and `C`. Divergence is more or less abs(substract(...)).
3. It could be also divergence = bitcount(xor) that tries to find most similar bits. then B and C is most coherent.

3 doesn't cope well with "it is free to ignore an arbitrary number of least
significant bits", but 1 and 2 fit well into definition.

I'm mentioning that because we discuss this feature only as reordering and forget that based on the ordering, the system needs to extract waves to execute.
Without understanding what is the actual "distance" between the hint codes its hard. It would be good to have that understanding on both sides of GPU.

Author: I believe arithmetic difference is the most flexible and intuitive definition for users. This approach allows a hint to incorporate both hierarchical aspects (with more important elements in MSBs) and ranges of bits that represent closeness along some axis. I agree it should be spelled out.

Reviewer 6: Agreed that it makes sense to add some details what we mean with a hierarchical hint, and giving an example.

However, I think the numerical value of the XOR makes more sense than arithmetic difference. We could also mention lexicographical sorting (even if the implementation doesn't explicitly sort) which many users should be familiar with. 

Consider:

```
hint1 = 0b1000
hint2 = 0b0111
hint3 = 0b1111
```

If the bits in the hint are intended to be used hierarchically (as per my understanding), then `hint2` is closer to `hint1` than to `hint3`. That is reflected in the numerical value of the XOR: `hint2 ^ hint1 = 0b111`, while `hint2 ^ hint3 = 0b1000`. However, using arithmetic difference, `hint2` would be closer to `hint3`.

Author: XOR prioritizes the strictly hierarchical use case. The problem with XOR is when a bit range represents similarity along some linear axis and we want to group based on closeness. 

Take for example the following X,Y hint pairs along some linear axis of similarity:
```
X  Y  Arith XOR
5  0  5     5
5  1  4     4
5  2  3     7
5  3  2     6
5  4  1     1
5  5  0     0
5  6  1     3
5  7  2     2
5  8  3     13
5  9  4     12
```

In this use case, arithmetic difference produces a desirable metric but not XOR.

Reviewer 6: I was interpreting the hints to have hierarchical semantics, but checking the spec again I couldn't find any explicit mentioning of that, so fine with me.

Again, good point that we should explain this.

Reviewer 5: I view it like in hint one encodes closeness in two metrics:
1. discreet space, any difference on given bits mean you are different, on this and that property. e.g.
   - you will use resource with this and that indices
   - shader will enter some if for some particular effect
2. some numeric that is arithmetic based on which you can calculate distance. the only example comes to my mind,  as @rasmusnv, mentions some linear things. I think the example that will be used most likely is parts of morton coded directions and origins.

now, the problem with arithmetic is that it potentially see the two hints with all bits different as the closest, so it breaks the 1. ... totally

on the other hand xor is succesfully used for checking morton builder.
even more crude, the closer code is considered the one that has lower first different bit (maybe I should put that on a list two, actually). so xor IMO should work great for 1. and good for 2. IMHO.

Author: I agree that there are some interesting aspects to specifying XOR between the two, but I think we need to consider something like this in the context of the entire system to see if it's reasonable.

Let's say a sort is performed that will put the hint values in descending order. Chunks of this sorted work is distributed to various execution units for dispatch. This should work for both arithmetic distance (it's an arithmetic sort) and XOR distance (the MSB has the highest weight on its order). The possible approximation with respect to the distance metric is at chunk boundaries.

As mentioned, this still leaves the question of how to prioritize dispatch groups within each execution unit when values are not a perfect match. This assumes that we are doing something sophisticated, rather than simply grabbing the next N sorted items.

In the earlier example, the three values would be ordered as:

```
B = 0b111111
A = 0b100000
C = 0b011111
```

Let's assume that we only have to group 2 items. We either pick A&B or A&C depending on if it's a XOR or arithmetic distance metric. Another choice is a strict **order** metric, which is actually the simplest in this case. That would indicate that we start with A&B.

Then consider the case where we have to group more work, say 8. Let's say that we have some duplicates (x-factor indicates the number of duplicates):

B = 0b111111 x 4
A = 0b100000 x 1
C = 0b011111 x 4

Would we be better served grouping B&C, B&A, A&C? I think this is not clear. What I'm trying to say is that the exact distance metric for dispatch grouping might be best left unspecified.

Maybe a suitable formulation would be that the system should in general try to dispatch hint values in arithmetic order (NOTE: not distance), but the exact mapping to waves is left unspecified.

Thoughts?

Reviewer 7: I would prefer if the expectation was we'd group A&B in your second example, based on the `XOR` distance. Given the grouping is allowed to throw away lower-order bits, it already implies higher order bits are more important, and it will be likely used to set "feature bits" anyways by ISVs because "arithmetic distance" between shaders is not really something you can calculate. But you can say "those two shaders are going to do X" and we want to group by similarity, so the more similar bits are set the more likely they are similar, with higher bits used to indicate "more expensive" things.

Undefined means it'll be defined by implementations, after all.

Reviewer 8: One thing to note here is that the XOR codes can be converted to arithmetic codes while preserving distance orderings at the cost of using a few more bits.  One such mapping is shown below.

![image](https://github.com/user-attachments/assets/cefe2da5-a2d1-4af4-aebc-91839893ea89)

In the a second example above we would have:

Label | XOR code | Distance preserving arithmetic code
---|---|---
B | 63 | 364
A | 32 | 243
C | 31 | 121

With this transform we can use an arithmetic distance, yet B&A remain closer than A&C.

Reviewer 7: I'm not convinced that it's easier for _applications_ to produce an arithmetic distance given applications will know features, not distances. I.e. a bit corresponds to a particular property, it's hard to imagine how an application would calculate an arithmetic distance from a set of feature bits. To me we got this backwards - now you'd have to take a "feature based metric" from the app, convert it to arithmetic distance using the reverse encoding, then unpack it to XOR again for sorting - what did we win? Would you mind posting the formula for both directions in this case?

Reviewer 8: For clarity, in the above the application isn't computing an arithmetic _distance_.  It's converting an xor _code_ into an arithmetic _code_ directly, such that the distance orderings are preserved.

The code conversion above reinterprets the xor code bits in trinary.  So where a bit would normally be interpreted to have value 2^i the conversion interprets it as 3^i.  So xor code 0b0101 (aka 5) becomes 3^2 + 3^0 = 10.  This inserts a gap between each new option bit sufficient to preserve distance orderings.

In application code this could be a change from `xor_code |= option_bit;` to `arith_code += option_value;`.  Of course we could imagine more complex scenarios than simply accumulating option constants.

I don't think it is possible to preserve arithmetic distance orderings using an xor code.  Maybe it can be done but it's not obvious to me how to do it.  So unfortunately I can't post the formula for the opposite direction.

>take a "feature based metric" ... what did we win?

Since there appear to be valid arguments for both arithmetic and xor distances what we won is having a single consistent code space (arithmetic) which the application can use to encode either arithmetic or xor distances.  The implementations can always sort and group by arithmetic distance and the application has the freedom to construct its codes to reflect either distance ordering.

Reviewer 5: The most important thing about arithmetic is that for me its not even good for arithmetic based HW if it drops bits.
So impl has to do rounding when doing dropping.

e.g.
original codes given by app:
A 101000000
B 100111111
C 101111111
if implementation drops 6 bits it sees:
A 101__________
B 100__________
C 101__________

For arithmetic the closest codes changes from A and B to A and C when bits are dropped.
for XOR it stays A and C.
Simply arithmetic is not stable for dropping of bits.

I wonder if we could pass literal with mask that says at least how the structure of hint looks like. for instance if hint by app is of structure:

```c
struct HintStructA {
  uchar featureA : 1;
  uchar featureB : 1;
  uchar LittleEnum : 2;
  uchar featureD : 1;
  uchar LittleArithCode : 3;
};
```
than ReorderThreads instead of having important bit count argument could have mask (compile time constant!) that marks on which bit each field starts. 
For above `hintStructA`: it would be `0b11101100`;

Dropping bits impl may be dropping whole fields. App can express its wish of understanding the hint including the ones we are discussing above:
e.g. 
arithmetic 8bit 0b10000000
xor            8bit 0b11111111

Author: The biggest disconnect for me is how these concepts map to the implementation. As mentioned earlier, arithmetic order and xor order is the same. The problem appears when we start talking about distances.

Distances are only useful if wave dispatch isn't strictly in order. We would have to extract the most similar hint pairs iteratively with a greedy approach. But that would leave the remaining work even less coherent, so it seems questionable as a global optimization. It's also unclear to me how such an algorithm would be parallelized and how hint distance would be weighted against thread occupancy.

If we specify the behavior as arithmetic or xor distance I'm afraid that apps will try strange mappings to convert between the two, wasting bits and being less efficient, for no gain on many implementations that don't care about distance.

Having a mask that indicates the start of each field is interesting and would allow the user to express intent exactly without introducing waste. If the mask is added by hand it may be confusing for developers, but having it generated from a struct may be a sweet spot. I think we would have to introduce a struct attribute for hints (similar to `[raypayload]`) which only allows bit fields.

I propose that we either go with @rdrabins' struct idea or that we don't specify anything about distances in the spec and instead talk about order. I have a slight preference to the latter since I don't see exactly how the information would be used in practice and since it adds complexity. The advantage of the struct is that it strictly gives more information about intent.

Reviewer 3: **Regarding "distance" metrics:**

We have some interesting discussions about different ways to compare hint values and compute distances, but as @rasmusnv [points out](https://github.com/microsoft/hlsl-specs/pull/277/files/bb617dfc02fd7d9ab67c532097a8f487b8fa0e98#r1900819781):
> The biggest disconnect for me is how these concepts map to the implementation.

Language specifying that higher bits of the hint value are generally considered more important than lower bits might help, but this likely needs qualifiers to temper expectations.

For instance, an implementation may combine buckets linearly to get to some size after sorting ([as suggested by rasmusnv](https://github.com/microsoft/hlsl-specs/pull/277/files/bb617dfc02fd7d9ab67c532097a8f487b8fa0e98#discussion_r1853742351)), which doesn't guarantee that groups of threads are reliably split along the highest-bit lines first, or even that the closest buckets by any "distance" metric are combined.

Example - selecting 16 threads from buckets sorted in ascending order:
```
A 0b0000 x 8
B 0b0100 x 8
C 0b0110 x 4
D 0b0111 x 4
```
A simple, greedy schedular might schedule A&B together before scheduling C&D, when globally it appears that B&C&D should be combined instead, since they have the same highest bit, and are even closest in arithmetic space.  If you use the concept of "distance", then in both XOR and in arithmetic space, B&C&D should be combined instead of A&B, yet that's not what this simple scheduler does.

I'm not saying this is how it *should* work, but I think this should be considered a valid implementation.

Given this, I suspect the language in the spec will need to remain somewhat vague to avoid reliance on any particular concept of "distance".

I suggest something like this:
> In general, more significant bits of the hint value are more important than less significant bits for determining coherency, though scheduling behavior is implementation specific.

---
**Regarding structured hints:**

This thread started on the basis of an insufficiently clear definition of how the hint values impact scheduling coherence.  If we cannot clearly define how more complex structured hint definitions map to behavior, the problem gets worse, not better.  The way I see it, a more complex hint definition sets up additional expectations for users that aren't likely to be met by implementations.

Author: @tex3d I committed your suggested change.

Reviewer 2: @rdrabins @rasmusnv resolving this thread if there are no other comments

not as obvious after a call to a hypothetical (fused)
`HitObject::ReorderAndInvoke`. Finally, tools can report live state across
`ReorderThread` and users can optimize live state across it. This is important
as live state across `ReorderThread` may be more expensive on some
architectures.

Some examples follow.

### Example: Common computations that rely on large raygeneration state

```C++
struct IorData
{
    float ior;
    uint id;
};

IorData iorList[16];
uint iorListSize = 0;

for( ... )
{
    HitObject hit = HitObject::TraceRay( ... );
    ReorderThread( hit );

    IorData newEntry = LoadIorDataFromHit( hit );
    bool enter = hit.GetHitKind() == HIT_KIND_TRIANGLE_FRONT_FACE;

    payload.ior = UpdateIorList( newEntry, enter, iorList, iorListSize );

    HitObject::Invoke( hit, payload );
}
```

### Example: Do common computations with hit-coherence

```C++
hit = HitObject::TraceRay( ... );
ReorderThread( hit );

payload.giData = GlobalIlluminationCacheLookup( hit );

HitObject::Invoke( hit, payload );
```

### Example: Same surface shader but different behavior in the raygeneration shader

```C++
// Primary ray wants perfect shadows and does not need to reorder as it is

> // coherent enough.

Reviewer 5: We'd appreciate using language that doesn't suggest that there is no reorder at trace ray.
It should be the expectation that impl should only minimize this. 
// Primary ray wants perfect shadows and does not need to reorder more than required and best performant for given implementation
// The domain of shadow rays is expected to be coherent enough.

Author: The latest commit includes clarifications like this.

duplicated in all closesthit shaders. Common code can simply be part of the
raygeneration shader and execute before closesthit shading. Second, simple
visibility rays no longer have to invoke hit shaders in order to access basic
information about the hit, such as the distance to the closest hit. Finally,
`HitObject` can be constructed from a `RayQuery`, which enables
`ReorderThread` and shader table-based closesthit and miss shading to be combined with
`RayQuery`.

The proposed extension to HLSL should be relatively straightforward to adopt by
current DXR implementations: `HitObject` merely decouples existing `TraceRay`
functionality into two distinct stages: the traversal stage and the shading
stage.
For SER's `ReorderThread`, the minimal allowed implementation is simply a
no-op, while implementations that already employ more sophisticated scheduling
strategies are likely able to reuse existing mechanisms to implement support
for SER. No DXR runtime changes are necessary, since the proposed extension to
the programming model is limited to HLSL and DXIL.

## Detailed Design

This section describes the HLSL additions for `HitObject` and `ReorderThread`
in detail.

> The canonical use of these features involve changing a `TraceRay` call to the
> following sequence:

Reviewer 5: we've spoken last time about this being equivalent functionally.
The current sentence "almost" says that. Could we have this more explicitly?
```suggestion
The canonical use of these features involve changing a `TraceRay` call to the
following sequence that is functionally equivalent:
```

    return A - B;
}
```

In this example, a number of different things could happen:
- If the implementation does not honor `ReorderThread` at all, the function
will most likely return zero, as the set of threads before and after the
conditional reorder would be the same.
- If the implementation reorders threads invoking `ReorderThread` but does not
replace them, B will likely be less than A for threads not invoking
`ReorderThread`, while the reordered threads will likely resume execution with
a newly formed full wave, thereby obtaining `A <= B`.
- If the implementation replaces threads in a wave, the threads not
participating in the reorder may possibly be joined by more threads than were
removed from the wave, again observing `A <= B`.

---

### Memory coherence and visibility

Due to the existence of non-coherent caches on most modern GPUs, an
application must take special care when communicating information across
reorder points through memory (UAVs).

This proposal includes a new coherence scope for reorder points. The scope
is limited to communication within the same dispatch index. Specifically,
if UAV stores or atomics are performed on one side of a reorder point, and

> on the other side the data is read via non-atomic UAV reads, the following

Author: It is mentioned that a shader transition is another reorder point so the same rules apply. Does that cover it? Do you have a relevant reference for Vulkan?

> on the other side the data is read via non-atomic UAV reads, the following

Reviewer 2: @rasmusnv AMD requested that the spec wording is more explicitly clarified for reordercoherent. No other outstandings, resolving thread.

is required:
- The UAV must be declared `[reordercoherent]`.
- The UAV writer must issue a `Barrier(UAV_MEMORY, REORDER_SCOPE)` between
the write and the reorder point.

When communicating both between threads and across reorder points, global coherency
can be utilized:
- The UAV must be declared `[globallycoherent]`.
- The UAV writer must issue a `DeviceMemoryBarrier` between the write and the

> reorder point.

Author: From my point of view, the `DeviceMemoryBarrier` call makes sure that all writes have completed before the reorder point gets to move the work to a different thread. When reading after the reorder point, there's no ordering problem. The only risk is that stale data is read from a local cache. That is mitigated by the `globallycoherent` attribute, causing the local cache to be bypassed.

> reorder point.

Reviewer 6: Let's separate the `reordercoherent` and `globallycoherent` discussions. I think `globallycoherent` in DXR semantics are a bit problematic because of the synchronization limitations, but that's not the focus of this change.

For `reordercoherent`, are you saying that you agree that not requiring a barrier between the reorder point and the load *does penalize* some implementations, but we accept that in favour of a simpler API that is closer to what e.g. WG require?

> reorder point.

Author: What I'm trying to say is that we are not trying to add acquire and release semantics to HLSL with this spec proposal.

What you are suggesting is calling `Barrier` both before and after the reorder point. First of all, note that you would have to have a barrier after each and every reorder point that wants to read the data after a single write since each reorder point will be a new thread. So a single write can result in a many required barriers, which may be difficult for the user. Second, there is no distinction between acquire and release with `Barrier` which means that all possible implementations are harmed (e.g., ordering + cache flush at both sides). Finally, it is not clear that there is any case where a high-frequency cache flush would be better than adopting the `globallycoherent` semantics of bypassing local caches, which all implementations already supports.

> reorder point.

Reviewer 6: > What I'm trying to say is that we are not trying to add acquire and release semantics to HLSL with this spec proposal.

Agreed.

> 
> What you are suggesting is calling `Barrier` both before and after the reorder point. First of all, note that you would have to have a barrier after each and every reorder point that wants to read the data after a single write since each reorder point will be a new thread. 

Nit: Let's not mix the thread terminology (as in the `globallycoherent` discussion) with the reorder terminology. We can say that at reorder points, `reordercoherent` synchronization is necessary. If we want to give it a name, maybe "execution locus" is a better term instead of "thread"?

> So a single write can result in a many required barriers, which may be difficult for the user. 

Agreed.

> Second, there is no distinction between acquire and release with `Barrier` which means that all possible implementations are harmed (e.g., ordering + cache flush at both sides). 

In the worst case, putting a `reordercoherent` barrier immediately after every reorder point (CallShader/TraceRay/ReportHit/HitObject ops / shader entry) should be equivalent to the current proposed behavior without the barrier on the reading side. Because it immediately follows a reorder point, it is clear that only reads can be affected, and if the implementation uses the bypass strategy nothing needs to be done. 
An implementation that doesn't use cache bypass would have to add exactly all these barriers, and the explicit one just makes that explicit.

> Finally, it is not clear that there is any case where a high-frequency cache flush would be better than adopting the `globallycoherent` semantics of bypassing local caches, which all implementations already supports.

Agreed that this is totally unclear in general, but I think when designing a specification that is supposed to be used for an extended lifetime, including on not yet existing hardware, one should be cautious to unnecessarily restrict the freedom of possible implementations.

To be clear, I am not demanding to change it to require the barrier on the reading side, I'm rather pointing out a potential issue. If there is consensus that the single-barrier solution is preferable, I'm totally fine with that.

> reorder point.

Reviewer 6: We re-discussed internally and are fine with the proposed solution of only requiring a barrier on the write side. It may penalize some implementations, but it's really a very special corner case where it makes a difference, and it's questionable whether it's worth the effort optimizing for that.

This variant of `ReorderThread` reorders threads based on a generic
user-provided hint. Similarity of hint values should indicate expected

> similarity of subsequent work being performed by threads. In general, more

Reviewer 3: I'm not sure how we resolve these issues without something vague, unless we prescribe only specific allowed interpretations.
- If they are always more important, that may imply XOR distance comparisons for grouping, but an arithmetic ordering may violate this expectation when grouping threads.
- They are more important in the sense that if bits are dropped, it will be some arbitrary number of low bits, making the lower bits less important.
- If they are not always more important, then what language do we use?

> similarity of subsequent work being performed by threads. In general, more

Reviewer 3: What about something like this:

> Thread grouping may be based on an arithmetic ordering of threads by hint value, or some other method.  Since an arbitrary number of lowest bits may be ignored, lower bits are less important than higher bits in the hint value.  However, thread grouping is not required to minimize any notion of distance between hint values within groups of threads, since this would greatly complicate thread grouping implementations.

> similarity of subsequent work being performed by threads. In general, more

Reviewer 3: Then again, if all parties are happy with @llvm-beanz's addition of `Specific scheduling behavior may vary by implementation` covering the divergence from any notion of distance, we could just accept that.

> similarity of subsequent work being performed by threads. In general, more

Reviewer 9: Is there any situation where `1 << 12` would be a "less important" hint than `1 << 2`?

These are hints, so we don't specifically dictate what the implementation _must_ do with the hint, but being clear about the ordering importance of bits seems like something we should all be able to agree on.

> similarity of subsequent work being performed by threads. In general, more

Reviewer 3: See my [last comment on the other thread](https://github.com/microsoft/hlsl-specs/pull/277#discussion_r1917520467) for an example of how threads with a different high bit could be grouped together while threads where the high bit is the same and only the lower bits are different are separated, when it could have grouped threads by the high bit value first instead.  If the high bit is always more important for grouping threads, the behavior observed from a simple greedy sorting scheduler may be unexpected.

In any case, I already said we could go with your suggested edit if everyone is satisfied with that.

> similarity of subsequent work being performed by threads. In general, more
> significant bits of the hint value are more important than less significant
> bits for determining coherency, though scheduling behavior is implementation
> specific. The resolution of

Reviewer 3: @llvm-beanz suggestion:

```suggestion
similarity of subsequent work being performed by threads. More significant
bits of the hint value are more important than less significant bits for
determining coherency. Specific scheduling behavior may vary by
implementation. The resolution of
```
